File: pages/scala/traversable.js


import Skb from 'components/skb-page'

export const id = "traversable"
const title = " Scala Traversable"

const date = "2021-01-13T17:00:07"

const scastieId = "ntaMRDzYQ3uMMgXbQZcECg"

const mainInfoBox = <>
  <p>Going a bit further in Functional Programming concepts with <i>Traversable</i></p>

<p>
We already saw a few concepts so far such as <code>Functor</code> and <code>Applicative</code>.<br>Let’s dig deeper !
</p>
</>

const detailedInfoBox = <>
  <p><i>Traversable</i> solves a specific problem. How to apply an operation to many inputs at once without leaving the “Box/Container/F” ?</p>

<p>
Without <i>Traversable</i>, we would have to perform something like this:
</p>

<pre>val input: List[Int] = List(1,2,3)
val operation: Int =&gt; F[Int] = ???

val output: List[F[Int]] = input.map( i =&gt; operation(i))
</pre>

<p>
But how do we manage to get <code>F[List[A]]</code> instead of <code>List[F[A]]</code> ? Traversable !
</p>

<p>
To implement <code>traverse</code> we re-use what we built in the previous episodes: <code>Functor</code> and <code>Applicative</code>. So we can chain the operation and wrap into <code>F</code>. 
</p>

<p>
With <code>traverse</code> which allows you to apply an operation to many element and stay inside <code>F</code>, we also get for free <code>sequence</code> which allow you to flip <code>List[F[A]]</code> inside out into <code>F[List[A]]</code>. You will see it in <code>Future.sequence</code> for instance that is very convenient to wait or combine many <code>Future</code> together.
</p>

<p>
As a note: <code>input.map(f).sequence</code> is the same as <code>input.traverse(f)</code>.
</p>
</>

const description = "Scala Knowledge Bits - Scala Traversable - Periodic exercise to learn bits of knowledge about Scala. Scala Traversable here."

const canonical_url = "https://leobenkel.com/2021/01/skb-scala-traversable/"

const prevUrl = "covariance"
const nextUrl = "typeclass"

const pageData = {
  id,
  title,
  date,
  scastieId,
  mainInfoBox,
  detailedInfoBox,
  description,
  canonical_url,
  prevUrl,
  nextUrl
}

export default function Page() {
  return <Skb lesson={pageData} />
}
 

File: snippets/traversable.scala
// Learn more about Scala on https://leobenkel.com

// Functor, Applicative and Traversable //

trait Functor[F[_]] {
  def map[A, B](fa: F[A])(f: A => B): F[B]
}

trait Applicative[F[_]] extends Functor[F] {
  def pure[A](a: A): F[A]

  def applicate[A, B](f: F[A => B])(a: F[A]): F[B]

  final override def map[A, B](a: F[A])(f: A => B): F[B] =
    applicate[A, B](pure[A => B](f))(a)
}

trait Traversable[F[_]] extends Applicative[F] {
  final def traverse[A, B](la: List[A])(f: A => F[B]): F[List[B]] = {
    la.foldLeft(pure(List.empty[B])) { case (acc, element) =>
      val fb: F[B] = f(???)
      val fCombine: F[B => List[B]] = map(???)((l: List[B]) => (a: B) => l :+ a)
      val output: F[List[B]] = applicate(???)(???)
      output
    }
  }

  // 'x => x' has a shortcut in Scala: 'identity' , try replacing it.
  final def sequence[A](lfa: List[F[A]]): F[List[A]] = traverse(???)(a => a)
}

//////////

// Example //

case class Box[A](v: A)

object Box extends Traversable[Box] {
  def pure[A](a: A) = Box[A](???)

  def applicate[A, B](ff: Box[A => B])(fa: Box[A]): Box[B] = Box(ff.v(???))
}

////////

// usages //
{ // traverse
  println("-Traverse-")
  val input: List[Int] = ???
  println(input)
  assert(input.length == 4)

  val output: Box[List[String]] = Box.traverse(input) {
    case 1               => ???
    case n if n % 2 == 0 => ???
    case a               => Box("x" * a)
  }
  println(output)
  assert(output.v.length == 4)
}

{ // sequence
  println("-Sequence-")
  val input: List[Box[Int]] = ???
  println(input)
  assert(input.length == 4)

  val output: Box[List[Int]] = Box.sequence(???)
  println(output)
  assert(output.v.length == 4)
}

////////

println(
  "Congratulations ! 'What would you do if you weren’t afraid?' - Sheryl Sandberg"
)
